Sources/GRPC/Compression/Zlib.swift

```diff
@@ -49,24 +49,27 @@ enum Zlib {
     /// - Parameter input: The complete data to be compressed.
     /// - Parameter output: The `ByteBuffer` into which the compressed message should be written.
     /// - Returns: The number of bytes written into the `output` buffer.
-    func deflate(_ input: inout Data, into output: inout ByteBuffer) throws -> Int {
+    func deflate(_ input: inout ByteBuffer, into output: inout ByteBuffer) throws -> Int {
       // Note: This is only valid because we always use Z_FINISH to flush.
       //
       // From the documentation:
       //   Note that it is possible for the compressed size to be larger than the value returned
       //   by deflateBound() if flush options other than Z_FINISH or Z_NO_FLUSH are used.
-      let upperBound = CGRPCZlib_deflateBound(&self.stream.zstream, UInt(input.count))
+      let upperBound = CGRPCZlib_deflateBound(&self.stream.zstream, UInt(input.readableBytes))
 
-      return try input.withUnsafeMutableBytes { inputPointer in
-        try output.writeWithUnsafeMutableBytes(minimumWritableBytes: Int(upperBound)) { outputPointer in
+      _ = input.readWithUnsafeMutableReadableBytes { inputPointer in
+        self.stream.prepareInputForDeflate(
+          inputBuffer: CGRPCZlib_castVoidToBytefPointer(inputPointer.baseAddress!),
+          inputBufferSize: inputPointer.count
+        )
+      }
```

Comment:
This isn't quite right: the input buffer pointer is now escaping the `input.readWithUnsafeMutableReadableBytes` closure (since it's assigned to the input buffer of `self.stream` which outlives the closure.

You should follow the pattern used for `inflate`; it should be something like:

```swift
fun deflate(...) -> Int {
  let upperBound =  ...
  
  return try input.readWithUnsafeMutableReadableBytes { inputPointer -> (Int, Int) in 
    self.stream.nextInputBuffer = ...
    self.stream.availableInputBytes = ... 

    defer {
      self.stream.nextInputBuffer = nil
      self.stream.availableInputBytes = 0
    }

    return try output.writeWithUnsafeMutableBytes(...) {
      ...
    }
}
```

Reply:
I'm still trying to fix this